Libraries/oneMKL/fourier_correlation/fcorr_2d_usm.cpp

```diff
@@ -6,151 +6,291 @@
+    auto sumReduction = sycl::reduction(temp, 0.0f, sycl::plus<float>());
+    cgh.parallel_for(sycl::range<2>{n_rows, n_cols},
+                     sumReduction,
+                     [=](sycl::id<2> idx, auto& sum) {
+      size_t row = idx[0];
+      size_t col = idx[1];
+      sum += img[row * col_stride + col] * img[row * col_stride + col];
+    });
+  }).wait();
+  const float frobenius_norm = sqrtf(temp[0]);
+  sycl::free(temp, Q);
+  return frobenius_norm;
+}
 
-    std::cout << std::endl << "First image:" << std::endl;
-    for (unsigned int r = 0; r < n_rows; r++) {
-        for (unsigned int c = 0; c < n_cols; c++) {
-            std::cout << img1[r * c_stride + c * r_stride] << " ";
-        }
-        std::cout << std::endl;
-    }
-    std::cout << std::endl << "Second image:" << std::endl;
-    for (unsigned int r = 0; r < n_rows; r++) {
-        for (unsigned int c = 0; c < n_cols; c++) {
-            std::cout << img2[r * c_stride + c * r_stride] << " ";
-        }
-        std::cout << std::endl;
+static sycl::event
+naive_cross_correlation(sycl::queue& Q,
+                        const unsigned& n_rows,
+                        const unsigned& n_cols,
+                        const unsigned& col_stride,
+                        const float* u,
+                        const float* v,
+                        float* w,
+                        const std::vector<sycl::event> deps = {}) {
+  // u, v and w must be USM allocations of (at least)
+  // ((n_rows - 1) * col_stride + n_cols - 1) float values (w must be writable)
+  if (!is_device_accessible(u, Q) ||
+      !is_device_accessible(v, Q) ||
+      !is_device_accessible(w, Q)) {
+    throw std::invalid_argument("Image arrays must be device-accessible");
+  }
+  sycl::event ev = Q.parallel_for(sycl::range<2>{n_rows, n_cols},
+                                  [=](sycl::id<2> idx) {
+    const size_t s = idx[0];
+    const size_t p = idx[1];
+    const size_t w_idx = s * col_stride + p;
+    w[w_idx] = 0.0f;
+    for (size_t j = 0; j < n_rows; j++) {
+      for (size_t k = 0; k < n_cols; k++) {
+        w[w_idx] += u[j * col_stride + k] *
+                    v[((j - s + n_rows) % n_rows) * col_stride
+                        + ((k - p + n_cols) % n_cols)];
+      }
     }
+  });
+  return ev;
+}
 
-    // Initialize FFT descriptor
-    oneapi::mkl::dft::descriptor<oneapi::mkl::dft::precision::SINGLE,
-                                 oneapi::mkl::dft::domain::REAL>
-        forward_plan({n_rows, n_cols});
-
-    // Data layout in real domain
-    std::int64_t real_layout[4] = {0, c_stride, 1};
-    forward_plan.set_value(oneapi::mkl::dft::config_param::INPUT_STRIDES,
-                           real_layout);
+int main(int argc, char **argv) {
+  int temp = (argc <= 1) ? 8 : std::stoi(argv[1]);
+  // n_rows >= 6 required for the arbitrary signals as defined herein
+  if (temp < 6)
+    throw std::invalid_argument("The first argument (n_rows) must be 6 or greater.");
```

Comment:
```suggestion
    throw std::invalid_argument("The number of rows of the image, chosen as first input of the program, must be 6 or greater.");
```

Reply:
Done in e776e50b2647a4e7d10808cd8a9c2617005766db